### PR TITLE
Fix for limited parallelism when the target is the same but has different args

### LINKED ISSUE #2377: WITH DOCKER is not executed in parallel in latest v0.6.28 (but it was in version v0.6.25)
Hi,

I have some targets meant for running tests projects in parallel that are using WITH DOCKER to run a Rabbimq service before running the tests. Using v0.6.25 I can see how  the test-executor instances are running fully in parallel loading the rabbitmq tar files in parallel etc. However when using v0.6.28 this is not the case anymore and the  test-executor instances  are running sequentially instead.  Is there a way to prevent this from happening?

```
parallel-testing:
    FROM +build
    WAIT
        FOR dir IN $(ls Tests/*Tests/*.csproj)
            COPY (+test-executor/TestResults --PROJECT=./$dir) ./TestResults
        END
        SAVE ARTIFACT ./TestResults testresults AS LOCAL earthly-artifacts/testresults
    END
```

```
   test-executor:
    FROM +build
    ARG PROJECT
    ARG ID
    COPY docker-compose-unit-tests-rabbitmq.yml .
    COPY coverlet.runsettings .

      WITH DOCKER \
     --compose docker-compose-unit-tests-rabbitmq.yml \
     --service rabbitmq
          RUN dotnet test ${PROJECT} --blame-hang-timeout 5m --no-build --collect:"XPlat Code Coverage" -p:CollectCoverage=true --settings coverlet.runsettings --logger trx --configuration $buildConfiguration --results-directory ./TestResults
      END
  
    SAVE ARTIFACT ./TestResults TestResults
```

### PR DESCRIPTION
**Pull request by vladaionescu**
Fixes #2377

Note that I was able to simplify the reproduction case - it is actually not directly related to `WITH DOCKER`. A difference in the implementation of the new `WITH DOCKER` command caused a pre-existing issue to surface.

The previous `VisitedCollection` implementation attempts to wait for ongoing targets with the same name to complete before comparing the inputs of the target. This has the advantage of being more precise with which ARGs actually influence the outcome of the targets, ignoring overriding ARGs that end up being unused within those targets. But it also has the disadvantage that all targets with the same name (but different overriding args) execute sequentially.

This new implementation simplifies this greatly, by computing the target input hash upfront based on all overriding args, without knowing if there are any args that will end up being unused. The result is that we are able to run all targets with the same name but different args in parallel. But it also has the disadvantage that in some cases we would create duplicated LLBs for some targets when certain overriding args are different but they are unused. Buildkit will generally de-duplicate the LLB in these cases, although it's possible that there might be edge cases if the LLB construction is not consistent.

### CODE REVIEW

File: states/visited2.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+		select {
+		case <-ctx.Done():
+			return nil, false, ctx.Err()
+		case <-sts.Done():
+			// Return existing.
+			return sts, true, nil
+		}
+	}
+	vc.visited[newKey] = newSts
+	vc.visitedList = append(vc.visitedList, newSts)
```
Comment (idelvall) [inline states/visited2.go:61]: What is the purpose of the list? Seems like we could remove it

File: states/visited2.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+		return nil, false, err
+	}
+	newKey, err := newSts.targetInput.Hash()
+	if err != nil {
+		return nil, false, err
+	}
+	vc.mu.Lock()
+	sts, found := vc.visited[newKey]
+	if found {
+		vc.mu.Unlock()
```
Comment (idelvall) [inline states/visited2.go:50]: Looks like `newSts.targetInput.Hash()` is not consistent with `newSts.targetInput.Equals()`, that is, we are getting cache hits (same hash), that are not considered equal. Not sure yet of the consequences but worth mentioning.

Changing to this test file
```
VERSION 0.7

FROM earthly/dind:alpine

parallel:
  BUILD +test-executor
  BUILD +test-executor --INDEX=2
  BUILD +test-executor --INDEX=3
  BUILD +test-executor --INDEX=4
  BUILD +test-executor --INDEX=5

test-executor:
  ARG INDEX=1
  RUN --no-cache for i in $(seq 1 3); do  \
        echo "### $INDEX"; \
        sleep 0.3; \
      done;
  IF true
    RUN true
  END
```
and adding here this code:
```
if !sts.targetInput.Equals(newSts.targetInput) {
			nb, _ := json.Marshal(newSts.targetInput)
			b, _ := json.Marshal(sts.targetInput)
			fmt.Println(fmt.Sprintf("%s %s\ncreated:%s\nfound:  %s", target.String(), strings.Join(overridingVars.Sorted(), " "), string(nb), string(b)))
		}
```
outputs:
```
./tests/with-docker-parallel+test-executor DOCKERHUB_MIRROR_AUTH_FROM_CLOUD_SECRETS INDEX
created:{"targetCanonical":"github.com/earthly/earthly/tests/with-docker-parallel:vlad/issue#2377+test-executor","buildArgs":[{"name":"DOCKERHUB_MIRROR_AUTH_FROM_CLOUD_SECRETS","constantValue":"true","defaultConstant":""},{"name":"INDEX","constantValue":"2","defaultConstant":""}],"platform":"linux/amd64","allowPrivileged":true}
found:  {"targetCanonical":"github.com/earthly/earthly/tests/with-docker-parallel:vlad/issue#2377+test-executor","buildArgs":[{"name":"DOCKERHUB_MIRROR_AUTH_FROM_CLOUD_SECRETS","constantValue":"true","defaultConstant":""},{"name":"INDEX","constantValue":"2","defaultConstant":"1"}],"platform":"linux/amd64","allowPrivileged":true}
./tests/with-docker-parallel+test-executor DOCKERHUB_MIRROR_AUTH_FROM_CLOUD_SECRETS INDEX
created:{"targetCanonical":"github.com/earthly/earthly/tests/with-docker-parallel:vlad/issue#2377+test-executor","buildArgs":[{"name":"DOCKERHUB_MIRROR_AUTH_FROM_CLOUD_SECRETS","constantValue":"true","defaultConstant":""},{"name":"INDEX","constantValue":"3","defaultConstant":""}],"platform":"linux/amd64","allowPrivileged":true}
found:  {"targetCanonical":"github.com/earthly/earthly/tests/with-docker-parallel:vlad/issue#2377+test-executor","buildArgs":[{"name":"DOCKERHUB_MIRROR_AUTH_FROM_CLOUD_SECRETS","constantValue":"true","defaultConstant":""},{"name":"INDEX","constantValue":"3","defaultConstant":"1"}],"platform":"linux/amd64","allowPrivileged":true}
./tests/with-docker-parallel+test-executor DOCKERHUB_MIRROR_AUTH_FROM_CLOUD_SECRETS INDEX
created:{"targetCanonical":"github.com/earthly/earthly/tests/with-docker-parallel:vlad/issue#2377+test-executor","buildArgs":[{"name":"DOCKERHUB_MIRROR_AUTH_FROM_CLOUD_SECRETS","constantValue":"true","defaultConstant":""},{"name":"INDEX","constantValue":"4","defaultConstant":""}],"platform":"linux/amd64","allowPrivileged":true}
found:  {"targetCanonical":"github.com/earthly/earthly/tests/with-docker-parallel:vlad/issue#2377+test-executor","buildArgs":[{"name":"DOCKERHUB_MIRROR_AUTH_FROM_CLOUD_SECRETS","constantValue":"true","defaultConstant":""},{"name":"INDEX","constantValue":"4","defaultConstant":"1"}],"platform":"linux/amd64","allowPrivileged":true}
./t
```

File: states/visited2.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+// All returns all visited items.
+func (vc *VisitedCollection2) All() []*SingleTarget {
+	vc.mu.Lock()
+	defer vc.mu.Unlock()
+	return append([]*SingleTarget{}, vc.visitedList...)
+}
+
+// Add adds a target to the collection, if it hasn't yet been visited. The returned sts is
+// either the previously visited one or a brand new one.
+// This function blocks if there is a visited target that is still running.
```
Comment (idelvall) [inline states/visited2.go:36]: >// This function blocks if there is a visited target that is still running.

This doesn't seem accurate, since it only blocks if there is cache hit to a running target?

File: states/visited2.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+import (
+	"context"
+	"sync"
+
+	"github.com/earthly/earthly/domain"
+	"github.com/earthly/earthly/util/platutil"
+	"github.com/earthly/earthly/variables"
+)
+
+// VisitedCollections is a collection of visited targets.
```
Comment (idelvall) [inline states/visited2.go:12]:
```suggestion
// VisitedCollections2 is a collection of visited targets.
```

File: states/visited2.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+type VisitedCollection2 struct {
+	mu      sync.Mutex
+	visited map[string]*SingleTarget // targetInputHash -> sts
+	// Same collection as above, but as a list, to make the ordering consistent.
+	visitedList []*SingleTarget
+}
+
+// NewVisitedCollection returns a collection of visited targets.
```
Comment (idelvall) [inline states/visited2.go:20]:
```suggestion
// NewVisitedCollection2 returns a collection of visited targets.
```

**Comment by idelvall** [comment]
PR for feature flag: https://github.com/earthly/earthly/pull/3407

File: states/visited2.go (hunk 3 above)
Comment (vladaionescu) [reply states/visited2.go:36]: Yes, correct - it only blocks if the target input hash matches.

File: states/visited2.go (hunk 2 above)
Comment (vladaionescu) [reply states/visited2.go:50]: This should be ok. The only difference is the `DefaultValue` field, which we want to ignore anyway in the new implementation.

File: states/visited2.go (hunk 3 above)
Comment (idelvall) [reply states/visited2.go]: What is the rationale behind this behavior?  
I mean why the first time visited it doesn't block but successive calls do?

**Comment by idelvall** [comment]
Will polish tests next

File: states/visited2.go (hunk 2 above)
Comment (idelvall) [reply states/visited2.go]: I would rather make `hash` and `equals` implementations consistent. 
What makes more sense, making `hash` dealing with the defaultConstant or `equals` ignoring it?

Comment (vladaionescu) [reply states/visited2.go]: Let me check if it's safe to drop it from equals - we really need hash to ignore it as part of this.

Comment (vladaionescu) [reply states/visited2.go]: Ok, I made these consistent now.

**Comment by idelvall** [comment]
Add tests PR: https://github.com/earthly/earthly/pull/3413

**Comment by idelvall** [comment]
We should run all the tests with this new visitor implementation, to be sure we didn't introduce any regressions. Notice that right now, the feature flag is disabled by default and hence the new visitor implementation is only used in the new test added

**Comment by alexcb** [comment]
> We should run all the tests with this new visitor implementation, to be sure we didn't introduce any regressions. Notice that right now, the feature flag is disabled by default and hence the new visitor implementation is only used in the new test added

we do something similar with the global wait/end code under: https://github.com/earthly/earthly/blob/2a9f2b10beec627d7a43238a9adb6516a833416d/.github/workflows/reusable-wait-block-target.yml#L58-L61

perhaps a similar approach could be done?

**Comment by vladaionescu** [comment]
> perhaps a similar approach could be done?

Would that add too much overhead to our CI stack? I think it's safe enough to let it go without extensive testing until we flip the flag in 0.8. I can see why this is necessary for the wait block though.

**Comment by alexcb** [comment]
> > perhaps a similar approach could be done?
> 
> Would that add too much overhead to our CI stack? I think it's safe enough to let it go without extensive testing until we flip the flag in 0.8. I can see why this is necessary for the wait block though.

You could always limit it to a target like `+test-quick`

**Comment by vladaionescu** [comment]
> You could always limit it to a target like +test-quick

I ended up enabling via `.earthly_version_flag_overrides`

File: states/visited2.go (hunk 3 above)
Comment (vladaionescu) [reply states/visited2.go]: I ended up moving it out, so that I can use the hash for infinite recursion checking before waiting.

File: Earthfile [hunk 6]
```diff
@@ -466,7 +466,7 @@ earthly-docker:
 # Otherwise, it will attempt to login to the docker hub mirror using the provided username and password
 earthly-integration-test-base:
     FROM +earthly-docker
-    RUN apk update && apk add pcre-tools curl python3 bash perl findutils expect yq
+    RUN apk update && apk add pcre-tools curl python3 bash perl findutils expect yq && apk add --upgrade sed
```
Comment (alexcb) [inline Earthfile:469]: I think it's worthwhile adding a comment why the `--upgrade` flag is needed here -- is it because the included version of sed is busybox, which doesn't work? (I had to run this target to come up with this guess)

**Review by alexcb — APPROVED** [review]
I _think_ it looks ok; I found the leaselock a bit hard to follow -- is there a reason we can't just use a global mutex instead for LOCALLY?

**Comment by vladaionescu** [comment]
> I think it looks ok; I found the leaselock a bit hard to follow -- is there a reason we can't just use a global mutex instead for LOCALLY?

The complexity is unfortunately necessary, due to this case:

```Earthly
target1:
  LOCALLY
  BUILD +target2
  ...

target2:
  LOCALLY
  ...
```

You need some sort to re-entrant lock. The leaselock is something of that sort.

**Comment by vladaionescu** [comment]
I ended up removing the locally lock as I noticed some dead-locking. I'll separate into a different PR since it's not related to these changes anyway (it was a pre-existing bug).

File: Earthfile (hunk 6 above)
Comment (idelvall) [reply Earthfile:469]: Yes, that is exactly the reason

**Review by alexcb — APPROVED** [review]

**Review by idelvall — APPROVED** [review]
LGTM
